### PR TITLE
Feat/initial func

### PR DESCRIPTION
**Pull request by c-blomquist**

### CODE REVIEW

**Comment by nrwinner** [comment]
This is pretty good! A few things I noticed during testing -
1. Executing a command that does not exist (`node index.js cheese`), for example, prints no output, should ideally communicate to the user that the command could not be found
2. Errors seem to print multiple times and, in some cases, dump raw output to the user (see screenshots)
Screenshot 2023-12-21 at 9 58 43 AM

Screenshot 2023-12-21 at 10 03 14 AM

Screenshot 2023-12-21 at 10 05 57 AM

Screenshot 2023-12-21 at 10 07 23 AM

File: databaseFunctions.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
+const { db, end } = require("./database.js");
+
+// Takes in task name for creation and outputs the id of the created task
```
Comment (nrwinner) [inline databaseFunctions.js:3]: Can we convert this comment to a properly-formatted [JSDoc](https://jsdoc.app/about-getting-started) comment?

File: databaseFunctions.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+async function addTask(taskName) {
+  try {
+    const query = `INSERT INTO task (name, completed, created_date)
+                        VALUES ($1, false, NOW())
+                        RETURNING id;`;
+
+    const res = await db.one(query, [taskName]);
```
Comment (nrwinner) [inline databaseFunctions.js:10]: Thank you for sanitizing input here :heart:

File: databaseFunctions.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    return res;
+  } catch (err) {
+    console.error("Error adding in task: " + err);
```
Comment (nrwinner) [inline databaseFunctions.js:13]: Since this is visible to the user, we probably want a different message here. The doesn't make a lot of sense 🤔

File: databaseFunctions.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  } finally {
+    end();
+  }
+}
+
+async function getTasks() {
+  try {
+    const query = `SELECT id, name, completed 
+                        FROM task;`;
+    const res = await db.manyOrNone(query);
```
Comment (nrwinner) [inline databaseFunctions.js:24]: Since we're not doing anything with `res` before returning, could we directly return the result from `db.manyOrNone`? Do we even need async/await here? This applies to all of the queries here, not just this one.

File: databaseFunctions.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    end();
+  }
+}
+
+async function getTask(taskID) {
+  try {
+    const query = `SELECT id, completed
+                        FROM task
+                        WHERE id = $1;`;
+    const res = await db.one(query, [taskID]);
```
Comment (nrwinner) [inline databaseFunctions.js:39]: See the comment about returning results directly and avoiding unnecessary async/await.

File: databaseFunctions.js (hunk 2 above)
Comment (nrwinner) [inline databaseFunctions.js:10]: See the comment about returning results directly and avoiding unnecessary async/await.

File: databaseFunctions.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    if (task.completed == true) {
+      console.error("Task is already completed!");
+      throw new Error("Task is already completed!");
+    }
+    const query = `UPDATE task
+                        SET completed = true, completed_date = NOW()
+                        WHERE id = $1
+                        RETURNING name, completed, completed_date;`;
+    const res = await db.one(query, [taskID]);
+    return res;
```
Comment (nrwinner) [inline databaseFunctions.js:59]: See the comment about returning results directly and avoiding unnecessary async/await.

File: databaseFunctions.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  } catch (err) {
+    console.error("Error being able to complete task: " + err);
```
Comment (nrwinner) [inline databaseFunctions.js:61]: Maybe just...
```suggestion
    console.error("Error completing task: " + err);
```

File: index.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    .catch((err) => {
+      console.error("Error with tasks returned: " + err);
+    });
+}
+
+async function createTask(taskName) {
+  if (!taskName) {
+    console.error("Please enter a task name to create.");
+  }
+  if (typeof taskName !== "string") {
```
Comment (nrwinner) [inline index.js:29]: I love this check, but I wonder if this should be an else/if rather than just an if - if the previous check succeeds (no task name is provided), do we need to check if it's a string?

File: index.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        console.error("Error with adding a task: " + err);
+      });
+  }
+}
+
+async function completeTask(taskID) {
+  if (!taskID) {
+    console.error("Please enter a task to be marked as complete.");
+  }
+  if (!/^\d+$/.test(taskID)) {
```
Comment (nrwinner) [inline index.js:47]: Similar thought - should this be an else/if?

File: index.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      })
+      .catch((err) => {
+        console.error("Error with adding a task: " + err);
+      });
+  }
+}
+
+async function completeTask(taskID) {
+  if (!taskID) {
+    console.error("Please enter a task to be marked as complete.");
```
Comment (nrwinner) [inline index.js:45]: We should probably specify a task ID here instead of just saying a task
```suggestion
    console.error("Please enter the ID of a task to be marked as complete.");
```

File: index.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  if (!/^\d+$/.test(taskID)) {
+    console.error("Please enter the task ID number that you want to complete.");
+  } else {
+    await db
+      .completeTask(taskID)
+      .then((res) => {
+        console.log(`Task ${res.name} completed on: ${res.completed_date}.`);
+      })
+      .catch((err) => {
+        console.error("Error with completing a task: " + err);
```
Comment (nrwinner) [inline index.js:56]: Hmm maybe reword this?

File: index.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+// create a new task, example: node index.js create "This is my new task"
+if (process.argv[2] === "create") {
+  createTask(process.argv[3]);
+}
+
+// complete a new task, example: node index.js complete 1
+if (process.argv[2] === "complete") {
+  completeTask(process.argv[3]);
+}
```
Comment (nrwinner) [inline index.js:78]: These could all be else/ifs right? If no command was inputted, could `argv[2]` ever be equal to "list"?

**Review by nrwinner — CHANGES_REQUESTED** [review]

**Comment by nrwinner** [comment]
Could you update the README file to include information about how to start the project? Ideally, we'd want a fresh developer to clone and configure their local environment without much oversight. This does not extend to environment variables, so they'll still need to grab our specific values. We should at least tell them that step exists and to reach out to someone for the values.

File: databaseFunctions.js (hunk 7 above)
Comment (c-blomquist) [reply databaseFunctions.js]: Changed error handling so errors do not output twice so this output is no longer needed.

File: databaseFunctions.js (hunk 3 above)
Comment (c-blomquist) [reply databaseFunctions.js]: Changed error handling so error messages only display once, so no longer applicable.

**Comment by c-blomquist** [comment]
@nrwinner 

I am not entirely sure how or what should be changed for the dumping raw outputs to the user, currently I am just returning whatever error is received and outputting that out to the user.

> 2. Errors seem to print multiple times and, in some cases, dump raw output to the user (see screenshots)
> 
> Screenshot 2023-12-21 at 10 03 14 AM Screenshot 2023-12-21 at 10 05 57 AM

**Comment by nrwinner** [comment]
> @nrwinner
> 
> I am not entirely sure how or what should be changed for the dumping raw outputs to the user, currently I am just returning whatever error is received and outputting that out to the user.
> 
> > 2. Errors seem to print multiple times and, in some cases, dump raw output to the user (see screenshots)
> > 
> > Screenshot 2023-12-21 at 10 03 14 AM Screenshot 2023-12-21 at 10 05 57 AM

Yeah on second thought, I'm not gonna block the PR for this. Don't think this will affect most of our users, but let's groom a ticket to dig into this further at a later date.

**Review by nrwinner — APPROVED** [review]
